core/client-entry.ts

```diff
@@ -63,8 +63,22 @@ const invokeClientEntry = async () => {
     })
   }
   router.onReady(async () => {
+    // check if app can be mounted
+    const canBeMounted = () => RouterManager.isRouteDispatched() && // route is dispatched
+      !(router as any).history.pending && // there is no pending in router history
+      !(app as any)._isMounted // it's not mounted before
+
+    if (canBeMounted()) {
+      app.$mount('#app')
+    }
     router.beforeResolve((to, from, next) => {
-      if (!from.name) return next() // do not resolve asyncData on server render - already been done
+      if (!from.name) {
+        next()
+        if (canBeMounted()) {
```

Comment:
shouldn't this be executed before next()?

Reply:
Problem is that vue-router doesn't reveal when it resolve async components in beforeResolve. So I check when `this.pending` is falsy, because it gets null just after enter events are finished ([here](https://github.com/vuejs/vue-router/blob/dev/src/history/base.js#L190)). In this [line](https://github.com/vuejs/vue-router/blob/dev/src/history/base.js#L185) it adds `this.router.resolveHooks` (beforeResolve) to other enter events. So we need to wait just after beforeResolve is finished and then we have falsy `this.pending`. Best solution would be mounting app after `queue` ends ([here](https://github.com/vuejs/vue-router/blob/dev/src/history/base.js#L143)), but I couldn't find any flag to check this in beforeResolve.

Reply:
I had to keep app mounting in onReady callback because in production version on for example homepage there could be a situation that beforeResolve event is not triggered at all.